pandas/tools/merge.py

```diff
@@ -110,46 +215,221 @@ def ordered_merge(left, right, on=None, left_by=None, right_by=None,
+                                       check_duplicates=False)
     elif right_by is not None:
-        if not isinstance(right_by, (list, tuple)):
-            right_by = [right_by]
-        pieces = []
-        for key, ypiece in right.groupby(right_by):
-            merged = _merger(left, ypiece)
-            for k in right_by:
-                try:
-                    if k in merged:
-                        merged[k] = key
-                except:
-                    pass
-            pieces.append(merged)
-        return concat(pieces, ignore_index=True)
+        result, _ = _groupby_and_merge(right_by, on, right, left,
+                                       lambda x, y: _merger(y, x),
+                                       check_duplicates=False)
     else:
-        return _merger(left, right)
+        result = _merger(left, right)
+    return result
+
+
+def merge_asof(left, right, on=None,
+               left_on=None, right_on=None,
+               by=None,
+               suffixes=('_x', '_y'),
+               tolerance=None,
+               allow_exact_matches=True,
+               check_duplicates=True):
+    """Perform an asof merge. This is similar to a left-join except that we
+    match on nearest key rather than equal keys.
+
+    For each row in the left DataFrame, we select the last row in the right
+    DataFrame whose 'on' key is less than or equal to the left's key. Both
+    DataFrames must be sorted by the key.
+
+    Optionally perform group-wise merge. This searches for the nearest match
+    on the 'on' key within the same group according to 'by'.
+
+    .. versionadded 0.18.2
+
+    Parameters
+    ----------
+    left : DataFrame
+    right : DataFrame
+    on : label or list
+        Field names to join on. Must be found in both DataFrames.
+        The data MUST be ordered. Furthermore this must be numeric index,
+        typically a datetimelike or integer-index.
+    left_on : label or list, or array-like
+        Field names to join on in left DataFrame. Can be a vector or list of
+        vectors of the length of the DataFrame to use a particular vector as
+        the join key instead of columns
+    right_on : label or list, or array-like
+        Field names to join on in right DataFrame or vector/list of vectors per
+        left_on docs
+    by : column name or list of column names
+        Group both the left and right DataFrames by the group columns; perform
+        the merge operation on these pieces and recombine.
+    suffixes : 2-length sequence (tuple, list, ...)
+        Suffix to apply to overlapping column names in the left and right
+        side, respectively
+    tolerance : integer or Timedelta, optional, default None
+        select asof tolerance within this range; must be compatible
+        to the merge index.
+    allow_exact_matches : boolean, default True
+
+        - If True, allow matching the same 'on' value
+            (i.e. less-than-or-equal-to)
```

Comment:
Can you align the `(i.e.  ` with `If` (otherwise sphinx will complain about unexpected indentation). Same for the lines below (and for `check_duplicates`)